File: src-tauri/src/processing/spring.rs
//! Spring physics simulation for smooth cursor movement
//!
//! Implements a damped spring system that provides natural-feeling
//! cursor smoothing by simulating physical spring dynamics.

use crate::project::schema::SpringConfig;

/// 1D spring state tracking position and velocity
#[derive(Debug, Clone)]
pub struct SpringState {
    pub position: f64,
    pub velocity: f64,
}

impl SpringState {
    /// Create a new spring state at the given position with zero velocity
    pub fn new(initial: f64) -> Self {
        Self {
            position: initial,
            velocity: 0.0,
        }
    }

    /// Advance the spring simulation by dt seconds toward the target
    ///
    /// Uses the damped harmonic oscillator equation:
    /// F = -k * x - c * v
    /// where k = stiffness, c = damping, x = displacement, v = velocity
    pub fn step(&mut self, target: f64, config: &SpringConfig, dt: f64) {
        let displacement = self.position - target;
        let spring_force = -config.stiffness * displacement;
        let damping_force = -config.damping * self.velocity;
        let acceleration = (spring_force + damping_force) / config.mass;

        self.velocity += acceleration * dt;
        self.position += self.velocity * dt;
    }

    /// Check if spring has settled (velocity and displacement below threshold)
    pub fn is_settled(&self, target: f64, threshold: f64) -> bool {
        (self.position - target).abs() < threshold && self.velocity.abs() < threshold
    }
}

/// 2D spring for cursor position (X and Y axes)
#[derive(Debug, Clone)]
pub struct Spring2D {
    pub x: SpringState,
    pub y: SpringState,
}

impl Spring2D {
    /// Create a new 2D spring at the given position
    pub fn new(x: f64, y: f64) -> Self {
        Self {
            x: SpringState::new(x),
            y: SpringState::new(y),
        }
    }

    /// Advance both X and Y springs toward the target position
    pub fn step(&mut self, target_x: f64, target_y: f64, config: &SpringConfig, dt: f64) {
        self.x.step(target_x, config, dt);
        self.y.step(target_y, config, dt);
    }

    /// Get the current smoothed position
    pub fn position(&self) -> (f64, f64) {
        (self.x.position, self.y.position)
    }

    /// Reset the spring to a new position with zero velocity (for teleports)
    pub fn reset(&mut self, x: f64, y: f64) {
        self.x = SpringState::new(x);
        self.y = SpringState::new(y);
    }

    /// Check if both axes have settled
    pub fn is_settled(&self, target_x: f64, target_y: f64, threshold: f64) -> bool {
        self.x.is_settled(target_x, threshold) && self.y.is_settled(target_y, threshold)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn default_config() -> SpringConfig {
        SpringConfig {
            stiffness: 470.0,
            damping: 70.0,
            mass: 3.0,
        }
    }

    #[test]
    fn test_spring_approaches_target() {
        let config = default_config();
        let mut state = SpringState::new(0.0);

        // Step toward target=100 for 1 second at 60fps
        for _ in 0..60 {
            state.step(100.0, &config, 1.0 / 60.0);
        }

        // Should be close to target
        assert!(
            (state.position - 100.0).abs() < 5.0,
            "Position {} should be close to 100",
            state.position
        );
    }

    #[test]
    fn test_spring_no_overshoot_with_high_damping() {
        let config = SpringConfig {
            stiffness: 470.0,
            damping: 150.0, // Higher damping
            mass: 3.0,
        };
        let mut state = SpringState::new(0.0);
        let mut max_pos = 0.0f64;

        // Step toward target=100 for 2 seconds
        for _ in 0..120 {
            state.step(100.0, &config, 1.0 / 60.0);
            max_pos = max_pos.max(state.position);
        }

        // Should not overshoot significantly with high damping
        assert!(
            max_pos < 105.0,
            "Max position {} should not overshoot much",
            max_pos
        );
    }

    #[test]
    fn test_spring_2d() {
        let config = default_config();
        let mut spring = Spring2D::new(0.0, 0.0);

        // Step toward (100, 200) for 1 second
        for _ in 0..60 {
            spring.step(100.0, 200.0, &config, 1.0 / 60.0);
        }

        let (x, y) = spring.position();
        assert!((x - 100.0).abs() < 5.0, "X {} should be close to 100", x);
        assert!((y - 200.0).abs() < 5.0, "Y {} should be close to 200", y);
    }

    #[test]
    fn test_spring_reset() {
        let config = default_config();
        let mut spring = Spring2D::new(0.0, 0.0);

        // Move toward target
        for _ in 0..30 {
            spring.step(100.0, 100.0, &config, 1.0 / 60.0);
        }

        // Reset to new position
        spring.reset(500.0, 500.0);

        let (x, y) = spring.position();
        assert_eq!(x, 500.0);
        assert_eq!(y, 500.0);
        assert_eq!(spring.x.velocity, 0.0);
        assert_eq!(spring.y.velocity, 0.0);
    }
}

File: src/processing/spring.ts
/**
 * Spring physics simulation for smooth cursor movement
 *
 * Implements a damped spring system that provides natural-feeling
 * cursor smoothing by simulating physical spring dynamics.
 *
 * This is a TypeScript port of the Rust implementation for real-time
 * preview in the editor.
 */

export interface SpringConfig {
  stiffness: number;
  damping: number;
  mass: number;
}

// High stiffness + damping for responsive cursor with minimal lag
// This ensures the cursor reaches its target quickly while still
// having smooth motion (no overshoot)
export const DEFAULT_SPRING_CONFIG: SpringConfig = {
  stiffness: 800,
  damping: 80,
  mass: 1,
};

export interface SpringState {
  position: number;
  velocity: number;
}

/**
 * Create a new spring state at the given position with zero velocity
 */
export function createSpringState(initial: number): SpringState {
  return { position: initial, velocity: 0 };
}

/**
 * Advance the spring simulation by dt seconds toward the target
 *
 * Uses the damped harmonic oscillator equation:
 * F = -k * x - c * v
 * where k = stiffness, c = damping, x = displacement, v = velocity
 *
 * @returns New spring state (immutable update)
 */
export function stepSpring(
  state: SpringState,
  target: number,
  config: SpringConfig,
  dt: number,
): SpringState {
  const displacement = state.position - target;
  const springForce = -config.stiffness * displacement;
  const dampingForce = -config.damping * state.velocity;
  const acceleration = (springForce + dampingForce) / config.mass;

  const newVelocity = state.velocity + acceleration * dt;
  const newPosition = state.position + newVelocity * dt;

  return { position: newPosition, velocity: newVelocity };
}

/**
 * Check if spring has settled (velocity and displacement below threshold)
 */
export function isSpringSettled(
  state: SpringState,
  target: number,
  threshold: number = 0.1,
): boolean {
  return (
    Math.abs(state.position - target) < threshold &&
    Math.abs(state.velocity) < threshold
  );
}

export interface Spring2DState {
  x: SpringState;
  y: SpringState;
}

/**
 * Create a new 2D spring at the given position
 */
export function createSpring2D(x: number, y: number): Spring2DState {
  return {
    x: createSpringState(x),
    y: createSpringState(y),
  };
}

/**
 * Advance both X and Y springs toward the target position
 *
 * @returns New spring state (immutable update)
 */
export function stepSpring2D(
  state: Spring2DState,
  targetX: number,
  targetY: number,
  config: SpringConfig,
  dt: number,
): Spring2DState {
  return {
    x: stepSpring(state.x, targetX, config, dt),
    y: stepSpring(state.y, targetY, config, dt),
  };
}

/**
 * Get the current position from a 2D spring
 */
export function getSpring2DPosition(state: Spring2DState): {
  x: number;
  y: number;
} {
  return { x: state.x.position, y: state.y.position };
}

/**
 * Check if both axes have settled
 */
export function isSpring2DSettled(
  state: Spring2DState,
  targetX: number,
  targetY: number,
  threshold: number = 0.1,
): boolean {
  return (
    isSpringSettled(state.x, targetX, threshold) &&
    isSpringSettled(state.y, targetY, threshold)
  );
}
